azkaban-common/src/main/java/azkaban/storage/HdfsStorage.java

```diff
@@ -76,18 +83,19 @@ public String putProject(final ProjectStorageMetadata metadata, final File local
       }
       final Path targetPath = new Path(projectsPath,
           StorageUtils.getTargetProjectFilename(metadata.getProjectId(), metadata.getHash()));
-      if (this.hdfs.exists(targetPath)) {
-        log.info(
-            String.format("Duplicate Found: meta: %s path: %s", metadata, targetPath));
-        return getRelativeProjectPath(targetPath);
-      }
+      final Path tmpPath = new Path(projectsPath, TMP_PROJECT_UPLOAD_FILENAME);
 
       // Copy file to HDFS
       log.info(String.format("Creating project artifact: meta: %s path: %s", metadata, targetPath));
-      this.hdfs.copyFromLocalFile(new Path(localFile.getAbsolutePath()), targetPath);
+      this.hdfs.copyFromLocalFile(true, true, new Path(localFile.getAbsolutePath()), tmpPath);
+
+      // Rename the tmp file to the final file and overwrite the final file if it already exists
+      // (i.e. if the hash is the same).
+      this.hdfs.rename(tmpPath, targetPath, Options.Rename.OVERWRITE);
```

Comment:
Does the other `rename` (without options) fail if target exists?

Reply:
Yes it does.

Reply:
Thanks!